### ISSUE TITLE
HEAT perfomance on other datasets

### ISSUE THREAD

**Issue by sawyer7246** [woodfrog/heat#4]
Thank you for releasing your code!
I notice that you only compared its performance on two tasks: outdoor architecture reconstruction & floorplan reconstruction.
Do you try HEAT on other datasets, like YorkUrban and Wireframe?

**Comment by woodfrog** [comment]
No, we didn't try these datasets. 

Structured reconstruction emphasizes more on the global geometric patterns of a scene, while these two line segment detection datasets focus more on the semantic meaning of lines (e.g., textures). Compared to the two datasets you mentioned, the [3D wireframe dataset](https://openaccess.thecvf.com/content_ICCV_2019/papers/Zhou_Learning_to_Reconstruct_3D_Manhattan_Wireframes_From_a_Single_Image_ICCV_2019_paper.pdf) is closer to our focus. As they introduced their dataset in page-2, this dataset "totally excludes planar textural lines, such as the vertical textures of Big Ben in Figure 1. The so-defined W aims to capture global scene geometry instead of local textural details." We didn't run our method on 3D wireframe, but you might consider this dataset as an extra benchmark.

**Comment by Tianhongbo0128** [comment]
About the Choice of positional encoding in HEAT Supplementary Document,you try the learnable embedding as the positional encoding is worse the trigonometric encoding, how do you make it learned? Is that using the corner model's feats as the  positional encoding?

**Comment by woodfrog** [comment]
> About the Choice of positional encoding in HEAT Supplementary Document,you try the learnable embedding as the positional encoding is worse the trigonometric encoding, how do you make it learned? Is that using the corner model's feats as the positional encoding?

No, it's just using trainable embedding vectors to replace the frequency-based positional encoding. This can be implemented by pytorch's nn.Embedding. Suppose the image size is 256, the learnable embedding will just be a 256xD trainable matrix.

**Comment by sawyer7246** [comment]
> No, we didn't try these datasets.
> 
> Structured reconstruction emphasizes more on the global geometric patterns of a scene, while these two line segment detection datasets focus more on the semantic meaning of lines (e.g., textures). Compared to the two datasets you mentioned, the [3D wireframe dataset](https://openaccess.thecvf.com/content_ICCV_2019/papers/Zhou_Learning_to_Reconstruct_3D_Manhattan_Wireframes_From_a_Single_Image_ICCV_2019_paper.pdf) is closer to our focus. As they introduced their dataset in page-2, this dataset "totally excludes planar textural lines, such as the vertical textures of Big Ben in Figure 1. The so-defined W aims to capture global scene geometry instead of local textural details." We didn't run our method on 3D wireframe, but you might consider this dataset as an extra benchmark.

Cool! I believe I've got your point. 
Actually, I am still wondering about the traits that make HEAT more suitable for handling "the global geometric patterns of a scene", because  I find that the "point-pair as edge" modeling strategy of HEAT can also apply to ordinary line segment detection(LSD). Though the LSD may focus on the semantic meaning of lines,  it should benefit from the global geometric pattern as well.
Could you elaborate a little bit about that?

**Comment by woodfrog** [comment]
> > No, we didn't try these datasets.
> > Structured reconstruction emphasizes more on the global geometric patterns of a scene, while these two line segment detection datasets focus more on the semantic meaning of lines (e.g., textures). Compared to the two datasets you mentioned, the [3D wireframe dataset](https://openaccess.thecvf.com/content_ICCV_2019/papers/Zhou_Learning_to_Reconstruct_3D_Manhattan_Wireframes_From_a_Single_Image_ICCV_2019_paper.pdf) is closer to our focus. As they introduced their dataset in page-2, this dataset "totally excludes planar textural lines, such as the vertical textures of Big Ben in Figure 1. The so-defined W aims to capture global scene geometry instead of local textural details." We didn't run our method on 3D wireframe, but you might consider this dataset as an extra benchmark.
> 
> Cool! I believe I've got your point. Actually, I am still wondering about the traits that make HEAT more suitable for handling "the global geometric patterns of a scene", because I find that the "point-pair as edge" modeling strategy of HEAT can also apply to ordinary line segment detection(LSD). Though the LSD may focus on the semantic meaning of lines, it should benefit from the global geometric pattern as well. Could you elaborate a little bit about that?

By the term "global geometric patterns," I meant the geometry rules based on which human architects or designers create the structures. You can imagine CAD drawings. For example, there are many relational patterns in CAD models of buildings -- the edges usually form quadrangles in the facade of a building, the roof can sometimes contain trapezia and triangles, and the entire building must be a set of closed polygons, etc. These geometric patterns are more "global" than the texture patterns and are independent of the image information (Or in other words, the existence of an edge can sometimes be determined by checking the locations and relations of other edges, without even looking at an image).

**Comment by sawyer7246** [comment]
Hi, I encounter another problem when reproducing the outdoor architecture reconstruction on the original HAWP model. I only achieve 36.2(sAP-10) with the best 256x256 model at 30 epochs. And the best score on HEAT metric like:
> corners - precision: 0.839 recall: 0.834 f_score: 0.836
> edges - precision: 0.775 recall: 0.335 f_score: 0.468
> regions - precision: 0.730 recall: 0.511 f_score: 0.601

I also notice that the performance gets worse after 40 epochs. 

Thought I convert the outdoor dataset in HEAT to HAWP-like-format which contains *edges_negative*, *edges_positive* of each sample. I do not know why it comes so different from the performance mentioned in your paper(Table 1).

Would you provide me with more details?

Thanks a lot.

**Comment by woodfrog** [comment]
> Hi, I encounter another problem when reproducing the outdoor architecture reconstruction on the original HAWP model. I only achieve 36.2(sAP-10) with the best 256x256 model at 30 epochs. And the best score on HEAT metric like:
> 
> > corners - precision: 0.839 recall: 0.834 f_score: 0.836
> > edges - precision: 0.775 recall: 0.335 f_score: 0.468
> > regions - precision: 0.730 recall: 0.511 f_score: 0.601
> 
> I also notice that the performance gets worse after 40 epochs.
> 
> Thought I convert the outdoor dataset in HEAT to HAWP-like-format which contains _edges_negative_, _edges_positive_ of each sample. I do not know why it comes so different from the performance mentioned in your paper(Table 1).
> 
> Would you provide me with more details?
> 
> Thanks a lot.

I followed their original training schedule but added some post-processing to refine HAWP's raw outputs. I do not recall all the details, but the main step is to merge the nearby corners to make the output a valid planar graph. If you need the post-processed outputs for HAWP and LETR, I can share them via email.

**Comment by sawyer7246** [comment]
Great! My email: sawyer.s.lee@gmail.com

**Comment by sawyer7246** [comment]
Pls. share the processed outdoor dataset for HAWP, if you still have it...

**Comment by woodfrog** [comment]
> Pls. share the processed outdoor dataset for HAWP, if you still have it...

I sent you the output results. Sorry that I have cleaned up the processed dataset for HAWP. I think you can just follow their codebase to do the pre-processing.

**Comment by sawyer7246** [comment]
> > Pls. share the processed outdoor dataset for HAWP, if you still have it...
> 
> I sent you the output results. Sorry that I have cleaned up the processed dataset for HAWP. I think you can just follow their codebase to do the pre-processing.

Received.  Thank you!

Actually, I've tried the HAWP on its own training set(the wireframes). It turns out to be almost the same score as reported in their paper.

While I convert the outdoor dataset by following their annotation rule with no other changes, then retraining it. It fails to meet the score you reported by a large margin.

I wonder if there are any points that I missed. Like the epoch to choose, learning rate, or other hyperparameters modification?

Any details are really helpful to me!

**Comment by woodfrog** [comment]
> > > Pls. share the processed outdoor dataset for HAWP, if you still have it...
> > 
> > 
> > I sent you the output results. Sorry that I have cleaned up the processed dataset for HAWP. I think you can just follow their codebase to do the pre-processing.
> 
> Received. Thank you!
> 
> Actually, I've tried the HAWP on its own training set(the wireframes). It turns out to be almost the same score as reported in their paper.
> 
> While I convert the outdoor dataset by following their annotation rule with no other changes, then retraining it. It fails to meet the score you reported by a large margin.
> 
> I wonder if there are any points that I missed. Like the epoch to choose, learning rate, or other hyperparameters modification?
> 
> Any details are really helpful to me!

I think all important details for adapting the baselines to our tasks are provided in the Appendix of our paper. Increasing the number of training epochs is crucial for HAWP. According to what you observed, it seems that you don't do data augmentation in a right way.

**Comment by sawyer7246** [comment]
Fine... I may test it on other models. 
Would you email me the remaining outputs by other models in Table 1?
Thank you so much~
